docs/list.man
.\" RCSID @(#)$Id: list.man,v 1.3 2002/03/07 20:44:17 rk Exp $
.\" LIBDIR
.TH "LIST" "3rko" "31 October 1998"
.SH NAME
list \- generic list ``object'' for Std-C that can contain any sort
of data type or struct.

.SH SYNOPSIS

.nf
#include "librko.h"

defines the following structures and typedef's:

typedef struct list list;
typedef struct list_elem list_elem;

/* warning sizeof(list) does not necessarily give the correct memory size
 * since we make it "expandable" to handle a variable size tag
 */
struct list {
	list_elem *first;		/* head of list */
	list_elem *last;		/* last in list */
	int number;			/* number in list */
	int (*addfn)(void **, va_list);	/* user fn to add user data */
	int (*delfn)(void **, va_list);	/* user fn to del user data */
	char tag[1];			/* name tag for list */
/* followed by expanded memory allocation to contain rest of tag with
 * terminating NULL.  A list object must only be defined as "list *"
 * and set with list_ctor(TAG,INIT,ADDFN,DELFN);
 */
};

struct list_elem {
	list_elem *prev;		/* previous one in list */
	list_elem *next;		/* next one in list */
	void *object;			/* pointer to object of interest */
};

/* shield users from  certain internal details */
#define LIST_OBJECT(le)	((le)->object)
#define LIST_NEXT(le)	((le)->next)
#define LIST_PREV(le)	((le)->prev)

list *list_ctor(const char *tag,
	int (addfn)(void **, va_list), int (delfn)(void **, va_list));
int list_dtor(list **lst, char const *tag, ...);
int list_exists(list const *lst, const char *tag);
int list_sizeof(list const *lst, const char *tag);
int list_number(list const *lst, const char *tag);
list_elem *list_first(list const *lst, const char *tag);
list_elem *list_last(list const *lst, const char *tag);
int list_push(list *lst, char const *tag, ...);
int list_pop(list *lst, char const *tag, ...);
int list_unshift(list *lst, char const *tag, ...);
int list_shift(list *lst, char const *tag, ...);
.fi

.SH DESCRIPTION
.I list
could have easily been made a C++ class, but is confined to C for
maximum portability.
.I list
is designed for easy constant time adding or deleting elements
at the beginning or end of the list.
It uses user-defined functions to handle the actual insertion
or deletion
of user data in a list element, which just contains a (void *)
pointer to the user data.
To see how this is done, it's best to look at the example below.
.P
All the functions refer to a ``tag'', a simple ASCII string that
the library will compare with.  This ensures that the list object
is not refered to incorrectly.
.P
All the functions, except for the ``constructor'' or
accessor functions, return 0 if no error occured, else
return <0 if an error occured.
The
.I list_pop
and
.I list_shift
functions will return 1, if they're executed on an empty list.
(This is not considered an error, but something the user may
wish to know.)

.TP 15
.I list_ctor
construct the list object, registering a tag and user functions for
handling the user data.
.TP
.I list_dtor
destroy the list object.
.TP
.I list_exists
returns 0 if the list is not constructed.
.TP
.I list_sizeof
returns the particular sizeof the list structure, which depends on
the tag size, otherwise returns -1
if the list is not defined.
.TP
.I list_number
returns the number of elements in the list, otherwise returns -1
if the list is not defined.
.TP
.I list_first
returns a pointer to the first list_elem in the list object.
It's intended for creating sequential access through the list.
.TP
.I list_last
returns a pointer to the last list_elem in the list object.
It's intended for creating sequential access through the list.
.TP
.I list_push
add an element to the end of the list.  It has a variable length
argument, which is passed to the user-defined add-function,
to allow data to be entered in the argument list.
.TP
.I list_pop
delete an element at the end of the list.  It has a variable
length argument, which is passed to the user-define delete-function,
to allow user data handling.
.TP
.I list_unshift
add an element to the beginning of the list.  It has a variable length
argument, which is passed to the user-defined add-function,
to allow data to be entered in the argument list.
.TP
.I list_shift
delete an element at the end of the list.  It has a variable
length argument, which is passed to the user-define delete-function,
to allow user data handling.
.TP
.I LIST_OBJECT
a macro that ``returns'' a pointer to the object given a list_elem
.TP
.I LIST_NEXT
a macro that ``returns'' a pointer to the next list_elem
.TP
.I LIST_PREV
a macro that ``returns'' a pointer to the previous list_elem
.TP
.I addfn(void **, va_list)
the user function for adding data in a list element.
It must return 0 if successful.
It's advised to return a value less than -1 for errors, since
list_unshift and list_push will return -1 on error.
.TP
.I delfn(void **, va_list)
the user function for deleting data in a list element.
It must return 0 if successful.
It's advised to return a value less than -1 for errors, since
list_shift and list_pop will return -1 on error
(and 1 if the list is empty).

.SH EXAMPLES
This example just stores, prints, and remove a list of ``double''
values.
.nf
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>

#define RKOERROR	/* to enable use of librko error facility */
#include "librko.h"

/*----- functions to handle a single element of "double" data -----*/
int addflt(void **ptr, va_list ap) {
	double *iptr;
	if ((iptr = malloc(sizeof(double))) == (double *) NULL)
		return -1;
	*iptr = va_arg(ap, double);
	*ptr = (void *)iptr;
	return 0;
}

int delflt(void **ptr, va_list ap) {
	free(*ptr);
	*ptr = (void *) NULL;
	return 0;
}

int main() {
	list *listobj = (list *) NULL;
	const char tagflt[] = "FLOATS";
	list_elem *eptr;

/* add elements to flt list */
	if ((listobj = list_ctor(tagflt, addflt, delflt)) == (list *) NULL)
		rkoperror("main");
	if (list_unshift(listobj, tagflt, 3.3)) rkoperror("main");
	if (list_unshift(listobj, tagflt, 2.2)) rkoperror("main");
	if (list_unshift(listobj, tagflt, 1.1)) rkoperror("main");
	if (list_push(listobj, tagflt, 2.2)) rkoperror("main");
	if (list_push(listobj, tagflt, 1.1)) rkoperror("main");
	
	/* display contents of list */
	printf("number=%d sizeof=%d", list_number(listobj, tagflt),
		list_sizeof(listobj, tagflt));
	for (eptr = list_first(listobj,tagflt); eptr != (list_elem *) NULL;
	eptr = LIST_NEXT(eptr)) {
		printf(" %3.1f :", *(double *)LIST_OBJECT(eptr));
	}
	printf("\n");

	/* destroy the list */
	if (list_dtor(&listobj, tagflt)) rkoperror("main");

	return 0;
}
.fi
should display when executed:
.nf
number=5 sizeof=30 1.1 : 2.2 : 3.3 : 2.2 : 1.1 :
.fi
.P
A more complex example is to use a user-defined structure.  The full
program will not be shown, except for the few differences from the
above code.

.nf
typedef struct mix mix;
struct mix {
	int ival;
	double fval;
};
/*----- handle a single element of "struct" data -----*/
int addmix(void **ptr, va_list ap) {
	mix *iptr;
	if ((iptr = malloc(sizeof(mix))) == (mix *) NULL)
		return -1;
	iptr->ival = va_arg(ap, int);
	iptr->fval = va_arg(ap, double);
	*ptr = (void *)iptr;
	return 0;
}
int delmix(void **ptr, va_list ap) {
	free(*ptr);
	*ptr = (void *) NULL;
	return 0;
}
 ...
	if (list_unshift(listobj, tagmix, 1, 1.1)) rkoerror("main");
 ...
	printf(" %d | %3.1f :",
		(((mix *)LIST_OBJECT(eptr))->ival),
		(((mix *)LIST_OBJECT(eptr))->fval));
.fi

.SH SEE ALSO
malloc(3),free(3),uvec(3rko),rkoerror(3rko)

.SH AUTHOR
R.K.Owen,Ph.D.

.KEY WORDS


NEWS

librko-1.0.0	2002/10/21
* first release with a working configure, using automake/autoconf etc.
* added spawn3



list.c
static const char RCSID[]="@(#)$Id: list.c,v 1.4 1998/10/29 16:58:49 rk Exp $";
static const char AUTHOR[]="@(#)list 1.0 10/31/98 R.K.Owen,Ph.D.";
/* list.c -
 * This could have easily been made a C++ class, but is
 * confined to C for easy portability.
 *
 * a "list" is similar to the C++ STL deque, but implemented
 * using (void *), and where the user needs to create & pass
 * certain helper functions.
 */
/*
 *********************************************************************
 *
 *     This software is copyrighted by R.K.Owen,Ph.D. 1998
 *
 * The author, R.K.Owen, of this software is not liable for any
 * problems WHATSOEVER which may result from use  or  abuse  of
 * this software. The author, R.K.Owen, grants unlimited rights
 * to anyone who uses and modifies this  software  for  private
 * non-commercial  use  as  long  as  this copyright and rights
 * notice remains in this software and is made available to all
 * recipients of this software.
 *
 * last known email: rkowen@ckns.net
 *                   rk@owen.sj.ca.us
 *
 *********************************************************************
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include "librko.h"

/* warning sizeof(list) does not necessarily give the correct memory size
 * since we make it "expandable" to handle a variable size tag
 */
/* ---------------------------------------------------------------------- */
/* list_ctor - construct a "list" object
 * returns NULL if an error, else the memory location  if OK.
 * list_ctor will call list_init() to set things up.
 */
list *list_ctor(const char *tag,
int (addfn)(void **, va_list), int (delfn)(void **, va_list)) {

	list *lst = (list *) NULL;

	if (!(lst = (list *) malloc(sizeof(list)+sizeof(tag)))) {
#ifdef RKOERROR
		(void) rkocpyerror("list_ctor : list malloc error!");
		rkoerrno = RKOMEMERR;
#endif
		return lst;
	}
	/* copy in tag to structure and initialize pointers */
	lst->first = (void *) NULL;
	lst->last = (void *) NULL;
	lst->number = 0;
	lst->addfn = addfn;
	lst->delfn = delfn;
	(void) strcpy(lst->tag, tag);

	return lst;
}
/* ---------------------------------------------------------------------- */
/* list_dtor - destroy the list object (needs user delfn if not empty).
 */
int list_dtor(list **lst, char const *tag, ...) {
	int retval = 0;
	list_elem *eptr = (list_elem *) NULL;
	list_elem *nptr = (list_elem *) NULL;
	void *ptr = (void *) NULL;
	va_list args;

	if (*lst == (void *) NULL) return 0;	/* already nulled */
	if (!list_exists(*lst, tag)) {
#ifdef RKOERROR
		(void) rkopsterror("list_dtor : ");
#endif
		return -1;
	}

	if ((*lst)->first != (void *) NULL && (*lst)->last != (void *) NULL) {
		eptr = (*lst)->first;
		va_start(args, tag);	/* get ready to pass extra args */
		while ((eptr != (list_elem *) NULL)) {
			ptr = eptr->object;
			if ((retval = ((*lst)->delfn)(&ptr, args))) {
#ifdef RKOERROR
				(void) rkocpyerror(
					"list_dtor : user function error!");
				rkoerrno = RKOMEMERR;
#endif
				goto unwind;
			}
			nptr = eptr->next;
			free(eptr);
			eptr = nptr;
		}
	}
	(*lst)->first = (*lst)->last = (list_elem *) NULL;
	(*lst)->number = -1;
	(*lst)->tag[0] = '\0';
	free (*lst);
#ifdef RKOERROR
	rkoerrno = RKO_OK;
#endif
unwind: va_end(args);
	return retval;
}
/* ---------------------------------------------------------------------- */
/* accessor functions */
int list_exists(list const *lst, const char *tag) {
	int retval = 0;
	if (lst == (list *) NULL) {
#ifdef RKOERROR
		(void) rkocpyerror("list_exists : null pointer!");
		rkoerrno = RKOUSEERR;
#endif
		retval = 0;
	} else {
		if (strncmp(lst->tag,tag, strlen(tag) + 1)) {
#ifdef RKOERROR
			(void) rkocpyerror("list_exists : list doesn't exist!");
			rkoerrno = RKOUSEERR;
#endif
			retval = 0;
		} else {
#ifdef RKOERROR
			rkoerrno = RKO_OK;
#endif
			retval = 1;
		}
	}
	return retval;
}

int list_number(list const *lst, const char *tag) {
	if (list_exists(lst, tag)) {
#ifdef RKOERROR
		rkoerrno = RKO_OK;
#endif
		return lst->number;
	} else {
#ifdef RKOERROR
		(void) rkopsterror("list_number : ");
#endif
		return -1;
	}
}

int list_sizeof(list const *lst, const char *tag) {
	if (list_exists(lst, tag)) {
#ifdef RKOERROR
		rkoerrno = RKO_OK;
#endif
		return sizeof(*lst) + strlen(tag);
	} else {
#ifdef RKOERROR
		(void) rkopsterror("list_number : ");
#endif
		return -1;
	}
}

list_elem *list_first(list const *lst, const char *tag) {
	if (list_exists(lst, tag)) {
#ifdef RKOERROR
		rkoerrno = RKO_OK;
#endif
		return lst->first;
	} else {
#ifdef RKOERROR
		(void) rkopsterror("list_first : ");
#endif
		return (list_elem *) NULL;
	}
}

list_elem *list_last(list const *lst, const char *tag) {
	if (list_exists(lst, tag)) {
#ifdef RKOERROR
		rkoerrno = RKO_OK;
#endif
		return lst->last;
	} else {
#ifdef RKOERROR
		(void) rkopsterror("list_last : ");
#endif
		return (list_elem *) NULL;
	}
}

/* ---------------------------------------------------------------------- */
/* list_add - add 1 element to the end of list */
int list_add(list *lst, char const *tag, ...) {

	int retval = 0;
	void *ptr = (void *) NULL;
	list_elem *eptr = (list_elem *) NULL;
	va_list args;

	if (!list_exists(lst, tag)) {
#ifdef RKOERROR
		(void) rkopsterror("list_add : ");
#endif
		return -1;
	}

	va_start(args, tag);	/* get ready to pass extra args */
	if ((retval = (lst->addfn)(&ptr, args)) != 0) {
#ifdef RKOERROR
		(void) rkocpyerror("list_add : user function error!");
		rkoerrno = RKOMEMERR;
#endif
		goto unwind;
	}
	if ((eptr = (list_elem *)malloc(sizeof(list_elem)))
	== (list_elem *) NULL) {
#ifdef RKOERROR
		(void) rkocpyerror("list_add : malloc error!");
		rkoerrno = RKOMEMERR;
#endif
		(void) (lst->delfn)(&ptr, args);
		goto unwind;
	}
	eptr->object = ptr;
	eptr->prev = (list_elem *) NULL;
	eptr->next = (list_elem *) NULL;

	if (lst->first == (list_elem *) NULL) {
		lst->first = eptr;
	}
	if (lst->last != (list_elem *) NULL) {
		eptr->prev = lst->last;
		lst->last->next = eptr;
	}
	lst->last = eptr;
	lst->number++;
#ifdef RKOERROR
	rkoerrno = RKO_OK;
#endif
unwind: va_end(args);
	return retval;
}
/* ---------------------------------------------------------------------- */
/* list_del - delete off 1 element at the end of list */
int list_del(list *lst, char const *tag, ...) {
	int retval = 0;
	void *ptr = (void *) NULL;
	list_elem *eptr = (list_elem *) NULL;
	va_list args;

	if (!list_exists(lst, tag)) {
#ifdef RKOERROR
		(void) rkopsterror("list_del : ");
#endif
		return -1;
	}

	if (lst->last == (list_elem *) NULL) {		/* already empty */
		return 1;				/* not an error */
	}

	eptr = lst->last;
	ptr = eptr->object;

	va_start(args, tag);	/* get ready to pass extra args */
	if ((retval = (lst->delfn)(&ptr, args))) {
#ifdef RKOERROR
		(void) rkocpyerror("list_del : user function error!");
		rkoerrno = RKOMEMERR;
#endif
		goto unwind;
	}

	lst->last = eptr->prev;
	if (lst->last == (list_elem *) NULL) {
		lst->first = (list_elem *) NULL;
	} else {
		lst->last->next = (list_elem *) NULL;
	}
	lst->number--;
	free(eptr);
#ifdef RKOERROR
	rkoerrno = RKO_OK;
#endif
unwind: va_end(args);
	return retval;
}
/* ---------------------------------------------------------------------- */
/* list_push - add 1 element to the beginning of list */
int list_push(list *lst, char const *tag, ...) {

	int retval = 0;
	void *ptr = (void *) NULL;
	list_elem *eptr = (list_elem *) NULL;
	va_list args;

	if (!list_exists(lst, tag)) {
#ifdef RKOERROR
		(void) rkopsterror("list_push : ");
#endif
		return -1;
	}

	va_start(args, tag);	/* get ready to pass extra args */
	if ((retval = (lst->addfn)(&ptr, args)) != 0) {
#ifdef RKOERROR
		(void) rkocpyerror("list_push : user function error!");
		rkoerrno = RKOMEMERR;
#endif
		goto unwind;
	}
	if ((eptr = (list_elem *)malloc(sizeof(list_elem)))
	== (list_elem *) NULL) {
#ifdef RKOERROR
		(void) rkocpyerror("list_push : malloc error!");
		rkoerrno = RKOMEMERR;
#endif
		(void) (lst->delfn)(&ptr, args);
		goto unwind;
	}
	eptr->object = ptr;
	eptr->prev = (list_elem *) NULL;
	eptr->next = (list_elem *) NULL;

	if (lst->first != (list_elem *) NULL) {
		eptr->next = lst->first;
		lst->first->prev = eptr;
	}
	lst->first = eptr;
	if (lst->last == (list_elem *) NULL) {
		lst->last = eptr;
	}
	lst->number++;
#ifdef RKOERROR
	rkoerrno = RKO_OK;
#endif
unwind: va_end(args);
	return retval;
}
/* ---------------------------------------------------------------------- */
/* list_pop - delete off 1 element at the beginning of list */
int list_pop(list *lst, char const *tag, ...) {
	int retval = 0;
	void *ptr = (void *) NULL;
	list_elem *eptr = (list_elem *) NULL;
	va_list args;

	if (!list_exists(lst, tag)) {
#ifdef RKOERROR
		(void) rkopsterror("list_pop : ");
#endif
		return -1;
	}

	if (lst->first == (list_elem *) NULL) {		/* already empty */
		return 1;				/* not an error */
	}

	eptr = lst->first;
	ptr = eptr->object;

	va_start(args, tag);	/* get ready to pass extra args */
	if ((retval = (lst->delfn)(&ptr, args)) != 0) {
#ifdef RKOERROR
		(void) rkocpyerror("list_pop : user function error!");
		rkoerrno = RKOMEMERR;
#endif
		goto unwind;
	}

	lst->first = eptr->next;
	if (lst->first == (list_elem *) NULL) {
		lst->last = (list_elem *) NULL;
	} else {
		lst->first->prev = (list_elem *) NULL;
	}
	lst->number--;
	free(eptr);
#ifdef RKOERROR
	rkoerrno = RKO_OK;
#endif
unwind: va_end(args);
	return retval;
}
